libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialogManager.cs

```diff
@@ -0,0 +1,109 @@
+        /// <param name="botFrameworkAuthentication">A <see cref="BotFrameworkAuthentication"/> used to obtain a client for making calls to Bot Builder Skills.</param>
+        /// <param name="scopes">Custom <see cref="MemoryScope"/> implementations that extend the memory system.</param>
+        /// <param name="pathResolvers">Custom <see cref="IPathResolver"/> that add new resolvers path shortcuts to memory scopes.</param>
+        /// <param name="dialogs">Custom <see cref="Dialog"/> that will be added to the root DialogSet.</param>
+        /// <param name="logger">An <see cref="ILogger"/> instance.</param>
+        public AdaptiveDialogManager(
+            string adaptiveDialogId,
+            string languageGeneratorId,
+            string defaultLocale,
+            ResourceExplorer resourceExplorer,
+            ConversationState conversationState,
+            UserState userState,
+            SkillConversationIdFactoryBase skillConversationIdFactoryBase,
+            BotFrameworkAuthentication botFrameworkAuthentication,
+            IEnumerable<MemoryScope> scopes = default,
+            IEnumerable<IPathResolver> pathResolvers = default,
+            IEnumerable<Dialog> dialogs = default,
+            ILogger logger = null)
+        {
+            this._defaultLocale = defaultLocale ?? "en";
+            this._adaptiveDialogId = adaptiveDialogId ?? throw new ArgumentNullException(nameof(adaptiveDialogId));
+            this._botFrameworkAuthentication = botFrameworkAuthentication ?? throw new ArgumentNullException(nameof(botFrameworkAuthentication));
+            this._resourceExplorer = resourceExplorer ?? throw new ArgumentNullException(nameof(resourceExplorer));
+            this._logger = logger ?? NullLogger<AdaptiveDialogManager>.Instance;
+
+            this.InitialTurnState.Add(this._resourceExplorer);
+            this.InitialTurnState.Add(this._botFrameworkAuthentication);
+            this.InitialTurnState.Add(this._logger);
+            this.InitialTurnState.Add(conversationState ?? throw new ArgumentNullException(nameof(conversationState)));
+            this.InitialTurnState.Add(userState ?? throw new ArgumentNullException(nameof(userState)));
+            this.InitialTurnState.Add(skillConversationIdFactoryBase ?? throw new ArgumentNullException(nameof(skillConversationIdFactoryBase)));
+            this.InitialTurnState.Add(scopes ?? Enumerable.Empty<MemoryScope>());
+            this.InitialTurnState.Add(pathResolvers ?? Enumerable.Empty<IPathResolver>());
+            this.InitialTurnState.Add(_languageGeneratorManagers.GetOrAdd(_resourceExplorer, _ => new LanguageGeneratorManager(_resourceExplorer)));
+            this.InitialTurnState.Add(_resourceExplorer.TryGetResource(languageGeneratorId, out var resource) ? (LanguageGenerator)new ResourceMultiLanguageGenerator(languageGeneratorId) : new TemplateEngineLanguageGenerator());
```

Comment:
I'm not sure I understand this line TryGetResource returns a resource, but that value is never used, we still create a new instance of ResourceMultiLanguageGenerator and we cast it to LanguageGenerator. The else statement is also confusing because it does create a new instance of TemplateEngineLanguageGenerator which is a different class than the one returned by the TryGetResource and different than the one I would have created if TryGetResource would have been succesful.

Reply:
Actually, this is a bug. The thing that is getting registered is LangaugeGenerator() not the derived classes.   

If we actually used the code that we shipped it would look like this:

           this.UseLanguageGeneration(languageGeneratorId);  
           this.UseLanguagePolicy(new LanguagePolicy(defaultLocale));